### DISCUSSION TITLE
Differential Privacy on textual data?

### DISCUSSION THREAD

**Discussion by yuxinxu77** [opendp/opendp#719] [Differential Privacy (Theory/Practice)] [answered]
Hi,

Do you happen to know how to perform differential privacy on textual data? For example, I have a project where I use sentences as the independent variable and put them in a transformer to perform multi-class classification. Is it possible to somehow synthesize the data? Thinking off the top of my head, we may have ways to do this: not by synthesizing text, but by adding noise to the embedding vectors after texts are converted into word embeddings. But I'm not sure if this idea is on the right track or not.

And, if I'm understanding SmartNoise correctly, it can only synthesize tabular data, not textual data?

**Comment by Shoeboxam** [comment]
It depends on what you want!

If you want to privately release a private text classifier, then start with a language model that has been pre-trained on public data, and then fine-tune with Opacus or Tensorflow-Privacy on your private dataset.

If you want to privately release the classifications of a text classifier on private data, from a model trained on private data, then try to apply PATE.

If you want to privately release the classifications of a text classifier on private data, from a public model, then apply randomized response on each output category, or the exponential mechanism on each vector of output probabilities, via parallel composition.

**Reply by yuxinxu77** [reply]
I see. But there’s no widely known way to create Differential-Privacy-compliant synthetic textual data based on real data, correct? I’m assuming that SmartNoise cannot do this?

**Reply by Shoeboxam** [reply, marked as answer]
Not through SmartNoise, and I don't know of any bespoke DP generative text models. But there are a variety of non-DP generative text models. If you want synthetic text, then take one of those pre-trained non-DP models, and then fine-tune it privately with a DP deep learning library. Starting with a pre-trained model is necessary because you don't want to spend your privacy budget learning the language representation.

I am, however, skeptical that the generated DP synthetic text will be useful, though. There's no telling how well the fine-tuned model will capture the distribution of your private dataset, and you will lose record associations (the generated text will no longer be linked to individuals you'd like to classify). But it is impossible to know for sure without trying it.

What do you want to release? Depending on what you want to release, there may be a simpler, cheaper approach.

**Reply by Shoeboxam** [reply]
This paper seems to be consistent with that intuition: https://arxiv.org/abs/2210.14348 They are more optimistic than I was, though! ;)

**Comment by yuxinxu77** [comment]
Thank you!

**Comment by joshua-oss** [comment]
Note that we have released a library [1] for HuggingFace and GPT-2 based on the paper Michael referenced.  You can get OK results by fine-tuning GPT-2 with DP-SGD, but it's very small model compared to state of the art.  I think we are a ways away from being able to fine-tune the larger models like 3.5 and 4.0 with differential privacy, and even the high-performing "smaller" state of the art text generation models like LLaMA 15B would be prohibitively costly to fine-tune with DP-SGD.

[1] https://github.com/microsoft/dp-transformers
